octree_web_viewer/src/bin/points_web_viewer.rs

```diff
@@ -12,44 +12,121 @@
+        .conflicts_with("octree_path")
+        .requires_all(&["prefix", "suffix", "octree_id"])
+}
 /// HTTP web viewer for 3d points stored in OnDiskOctrees
 #[derive(StructOpt, Debug)]
-#[structopt(name = "points_web_viewer", about = "Visualizing points")]
-struct CommandlineArguments {
+#[structopt(
+    name = "points_web_viewer",
+    about = "Visualizing points",
+    raw(group = "group_inputs()")
+)]
+pub struct CommandLineArguments {
     /// The octree directory to serve, including a trailing slash.
-    octree_path: String,
+    /// this overrides the path_* options
+    #[structopt(name = "DIR", parse(from_os_str))]
+    octree_path: Option<PathBuf>,
     /// Port to listen on.
-    #[structopt(default_value = "5433", long = "--port")]
+    #[structopt(default_value = "5433", long = "port")]
     port: u16,
     /// IP string.
-    #[structopt(default_value = "127.0.0.1", long = "--ip")]
+    #[structopt(default_value = "127.0.0.1", long = "ip")]
     ip: String,
+    /// instead of DIR: specify path prefix for octree dir
+    #[structopt(long = "prefix", parse(from_os_str), group = "path_stubs")]
+    path_prefix: Option<PathBuf>,
+    /// Optional: suffix for subfolder of octree dir
+    #[structopt(long = "suffix", parse(from_os_str))]
+    path_suffix: Option<PathBuf>,
+    /// instead of DIR: specify path folder for octree dir
+    #[structopt(long = "octree_id", group = "path_stubs")]
+    octree_id: Option<String>,
+    /// Cache items
+    #[structopt(default_value = "20", long = "cache_items")]
+    cache_max: usize,
+}
+
+/// init app state with command arguments
+/// backward compatibilty is ensured
+pub fn state_from(args: CommandLineArguments) -> Result<AppState, PointsViewerError> {
+    let suffix = args.path_suffix.unwrap_or_else(|| PathBuf::from(""));
+
+    let app_state = match args.octree_path {
+        Some(octree_directory) => {
+            let prefix_opt = octree_directory.parent();
+            if suffix.to_string_lossy().is_empty() {
+                if let Some(prefix) = prefix_opt {
+                    let octree_id = octree_directory.strip_prefix(&prefix)?;
+                    AppState::new(args.cache_max, prefix, suffix, octree_id.to_str().unwrap())
+                } else {
+                    AppState::new(args.cache_max, PathBuf::new(), suffix, "/")
+                }
+            } else {
+                let mut components = octree_directory.components();
+                let mut prefix: PathBuf = PathBuf::new();
+                let mut tmp_octree_id = components.next();
+                while !suffix.as_path().eq(components.as_path()) {
+                    prefix.push(tmp_octree_id.unwrap());
+                    tmp_octree_id = components.next();
+                }
+
+                AppState::new(
+                    args.cache_max,
+                    prefix,
+                    suffix,
+                    (tmp_octree_id.unwrap().as_ref() as &Path).to_string_lossy(),
```

Comment:
Can we replace this with `tmp_octree_id.unwrap().to_string_lossy()`?

Reply:
unfortunately it is not possible
`no method named `to_string_lossy` found for type `std::path::Component<'_>` in the current scope
  --> octree_web_viewer/src/bin/points_web_viewer.rs:88:45
   |
88 |                     (tmp_octree_id.unwrap().to_string_lossy()),`